api/src/main/java/org/pmiops/workbench/api/DataSetController.java

```diff
@@ -253,16 +253,18 @@ public ResponseEntity<ReadOnlyNotebookResponse> previewExportToNotebook(
 
     JSONObject notebookFile = createNotebookObject(dataSetExportRequest.getKernelType());
 
-    dataSetService
-        .generateCodeCells(dataSetExportRequest, dbWorkspace)
-        .forEach(
-            cell -> notebookFile.getJSONArray("cells").put(createNotebookCodeCellWithString(cell)));
+    List<String> codeCells = dataSetService.generateCodeCells(dataSetExportRequest, dbWorkspace);
+
+    String rawCode = String.join(System.lineSeparator(), codeCells);
+    codeCells.forEach(
```

Comment:
I realize that this mutation of the `notebookFile` is already present in the existing code, but I expect there's a way we can do this immutably.

Not sure of the final step in that process, but something simple like `notebookFile = addCells(notebookFile, cells)` would help clarify what's going on here.  We could use this in `exportToNotebook()` too

Reply:
I am not sure what you mean by doing this immutably.

My understanding of what is happening here is that:
`JSONObject notebookFile = createNotebookObject(dataSetExportRequest.getKernelType());`
 this line makes a json representation of a notebook with an empty json array for cells.
 
 `    List<String> codeCells = dataSetService.generateCodeCells(dataSetExportRequest, dbWorkspace);` 
This line generates the code for each cell for the request

`    codeCells.forEach(
        cell -> notebookFile.getJSONArray("cells").put(createNotebookCodeCellWithString(cell)));`
This line takes the code, transforms it to notebook syntax, and then adds that cell to the notebookFile's cells array.

`            .html(
                notebooksService.convertJupyterNotebookToHtml(
                    notebookFile.toString().getBytes(StandardCharsets.UTF_8)))`
                    
This line takes the constructed json and converts it to html.

Reply:
this is optional, doesn't block.

I just prefer a coding style which creates a new object when making a modification rather than modifying existing objects.  We mostly stick this this, and it's easier to reason about.

Reply:
I think that I understand now. Thanks for explaining.

Reply:
I took a pass at this. What do you think?
https://github.com/all-of-us/workbench/pull/8248/commits/95a0b3a0e3aee546b2f97d15ff20c573963856d0